File: cpp/test_algorithms.cpp
#include "dsu_2pass.hpp"
#include "algorithms.hpp"
#include <iostream>
#include <vector>
#include <cassert>
#include <set>

bool test_simple_4_connected() {
    // Test image: 2 components
    // [1, 1, 0, 0]
    // [1, 0, 0, 0]
    // [0, 0, 1, 1]
    // [0, 0, 1, 1]
    int H = 4, W = 4;
    std::vector<uint8_t> img = {
        1, 1, 0, 0,
        1, 0, 0, 0,
        0, 0, 1, 1,
        0, 0, 1, 1
    };
    
    auto result_2pass = label_cc_2pass(img.data(), H, W, false);
    auto result_bfs = label_cc_bfs(img.data(), H, W, false);
    auto result_dfs = label_cc_dfs(img.data(), H, W, false);
    auto result_dsu = label_cc_dsu(img.data(), H, W, false);
    
    // Count unique labels (excluding 0)
    std::set<int32_t> labels_2pass, labels_bfs, labels_dfs, labels_dsu;
    for (int i = 0; i < H * W; ++i) {
        if (result_2pass[i] > 0) labels_2pass.insert(result_2pass[i]);
        if (result_bfs[i] > 0) labels_bfs.insert(result_bfs[i]);
        if (result_dfs[i] > 0) labels_dfs.insert(result_dfs[i]);
        if (result_dsu[i] > 0) labels_dsu.insert(result_dsu[i]);
    }
    
    assert(labels_2pass.size() == 2);
    assert(labels_bfs.size() == 2);
    assert(labels_dfs.size() == 2);
    assert(labels_dsu.size() == 2);
    
    return true;
}

bool test_simple_8_connected() {
    // Test image: diagonal pattern
    // [1, 0, 1]
    // [0, 1, 0]
    // [1, 0, 1]
    int H = 3, W = 3;
    std::vector<uint8_t> img = {
        1, 0, 1,
        0, 1, 0,
        1, 0, 1
    };
    
    // 8-connectivity: should be 1 component
    auto result_2pass_8 = label_cc_2pass(img.data(), H, W, true);
    auto result_bfs_8 = label_cc_bfs(img.data(), H, W, true);
    
    std::set<int32_t> labels_2pass_8, labels_bfs_8;
    for (int i = 0; i < H * W; ++i) {
        if (result_2pass_8[i] > 0) labels_2pass_8.insert(result_2pass_8[i]);
        if (result_bfs_8[i] > 0) labels_bfs_8.insert(result_bfs_8[i]);
    }
    
    assert(labels_2pass_8.size() == 1);
    assert(labels_bfs_8.size() == 1);
    
    // 4-connectivity: should be 5 components
    auto result_2pass_4 = label_cc_2pass(img.data(), H, W, false);
    std::set<int32_t> labels_2pass_4;
    for (int i = 0; i < H * W; ++i) {
        if (result_2pass_4[i] > 0) labels_2pass_4.insert(result_2pass_4[i]);
    }
    
    assert(labels_2pass_4.size() == 5);
    
    return true;
}

bool test_empty_image() {
    int H = 10, W = 10;
    std::vector<uint8_t> img(H * W, 0);
    
    auto result_2pass = label_cc_2pass(img.data(), H, W);
    auto result_bfs = label_cc_bfs(img.data(), H, W);
    
    for (int i = 0; i < H * W; ++i) {
        assert(result_2pass[i] == 0);
        assert(result_bfs[i] == 0);
    }
    
    return true;
}

bool test_full_foreground() {
    int H = 5, W = 5;
    std::vector<uint8_t> img(H * W, 1);
    
    auto result_2pass = label_cc_2pass(img.data(), H, W);
    auto result_bfs = label_cc_bfs(img.data(), H, W);
    
    std::set<int32_t> labels_2pass, labels_bfs;
    for (int i = 0; i < H * W; ++i) {
        if (result_2pass[i] > 0) labels_2pass.insert(result_2pass[i]);
        if (result_bfs[i] > 0) labels_bfs.insert(result_bfs[i]);
    }
    
    assert(labels_2pass.size() == 1);
    assert(labels_bfs.size() == 1);
    
    return true;
}

bool test_consistency() {
    // Random test image
    int H = 50, W = 50;
    std::vector<uint8_t> img(H * W);
    // Simple deterministic pattern
    for (int i = 0; i < H * W; ++i) {
        img[i] = (i % 7 < 2) ? 1 : 0;
    }
    
    auto result_2pass = label_cc_2pass(img.data(), H, W);
    auto result_bfs = label_cc_bfs(img.data(), H, W);
    auto result_dfs = label_cc_dfs(img.data(), H, W);
    auto result_dsu = label_cc_dsu(img.data(), H, W);
    
    std::set<int32_t> labels_2pass, labels_bfs, labels_dfs, labels_dsu;
    for (int i = 0; i < H * W; ++i) {
        if (result_2pass[i] > 0) labels_2pass.insert(result_2pass[i]);
        if (result_bfs[i] > 0) labels_bfs.insert(result_bfs[i]);
        if (result_dfs[i] > 0) labels_dfs.insert(result_dfs[i]);
        if (result_dsu[i] > 0) labels_dsu.insert(result_dsu[i]);
    }
    
    assert(labels_2pass.size() == labels_bfs.size());
    assert(labels_bfs.size() == labels_dfs.size());
    assert(labels_dfs.size() == labels_dsu.size());
    
    return true;
}

int main() {
    std::cout << "Running C++ tests...\n\n";
    
    try {
        if (test_simple_4_connected()) {
            std::cout << "✓ Simple 4-connected test passed\n";
        }
        if (test_simple_8_connected()) {
            std::cout << "✓ Simple 8-connected test passed\n";
        }
        if (test_empty_image()) {
            std::cout << "✓ Empty image test passed\n";
        }
        if (test_full_foreground()) {
            std::cout << "✓ Full foreground test passed\n";
        }
        if (test_consistency()) {
            std::cout << "✓ Consistency test passed\n";
        }
        
        std::cout << "\n✅ All tests passed!\n";
        return 0;
    } catch (const std::exception& e) {
        std::cerr << "❌ Test failed: " << e.what() << "\n";
        return 1;
    } catch (...) {
        std::cerr << "❌ Unknown error occurred\n";
        return 1;
    }
}

File: python/test_algorithms.py
#!/usr/bin/env python3
"""
Test suite for Connected Component Labeling algorithms
"""
import numpy as np
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent))

from dsu_2pass import label_cc_2pass
from algorithms import label_cc_bfs, label_cc_dfs, label_cc_dsu


def test_simple_4_connected():
    """Test simple 4-connected case"""
    img = np.array([
        [1, 1, 0, 0],
        [1, 0, 0, 0],
        [0, 0, 1, 1],
        [0, 0, 1, 1]
    ], dtype=np.uint8)
    
    result_2pass = label_cc_2pass(img, eight_connectivity=False)
    result_bfs = label_cc_bfs(img, eight_connectivity=False)
    result_dfs = label_cc_dfs(img, eight_connectivity=False)
    result_dsu = label_cc_dsu(img, eight_connectivity=False)
    
    # Should have 2 components
    assert len(np.unique(result_2pass[result_2pass > 0])) == 2, "2-pass failed"
    assert len(np.unique(result_bfs[result_bfs > 0])) == 2, "BFS failed"
    assert len(np.unique(result_dfs[result_dfs > 0])) == 2, "DFS failed"
    assert len(np.unique(result_dsu[result_dsu > 0])) == 2, "DSU failed"
    
    # All should have same number of components
    assert (np.unique(result_2pass[result_2pass > 0]).size == 
            np.unique(result_bfs[result_bfs > 0]).size), "Mismatch"
    
    print("✓ Simple 4-connected test passed")


def test_simple_8_connected():
    """Test simple 8-connected case"""
    img = np.array([
        [1, 0, 1],
        [0, 1, 0],
        [1, 0, 1]
    ], dtype=np.uint8)
    
    # With 8-connectivity, should be 1 component
    result_2pass = label_cc_2pass(img, eight_connectivity=True)
    result_bfs = label_cc_bfs(img, eight_connectivity=True)
    
    assert len(np.unique(result_2pass[result_2pass > 0])) == 1, "2-pass 8-conn failed"
    assert len(np.unique(result_bfs[result_bfs > 0])) == 1, "BFS 8-conn failed"
    
    # With 4-connectivity, should be 5 components
    result_2pass_4 = label_cc_2pass(img, eight_connectivity=False)
    assert len(np.unique(result_2pass_4[result_2pass_4 > 0])) == 5, "2-pass 4-conn failed"
    
    print("✓ Simple 8-connected test passed")


def test_empty_image():
    """Test empty (all background) image"""
    img = np.zeros((10, 10), dtype=np.uint8)
    
    result_2pass = label_cc_2pass(img)
    result_bfs = label_cc_bfs(img)
    
    assert np.all(result_2pass == 0), "2-pass empty failed"
    assert np.all(result_bfs == 0), "BFS empty failed"
    
    print("✓ Empty image test passed")


def test_full_foreground():
    """Test full foreground image"""
    img = np.ones((5, 5), dtype=np.uint8)
    
    result_2pass = label_cc_2pass(img)
    result_bfs = label_cc_bfs(img)
    
    # Should have exactly 1 component
    assert len(np.unique(result_2pass[result_2pass > 0])) == 1, "2-pass full failed"
    assert len(np.unique(result_bfs[result_bfs > 0])) == 1, "BFS full failed"
    
    print("✓ Full foreground test passed")


def test_consistency():
    """Test that all algorithms produce consistent results"""
    np.random.seed(42)
    img = (np.random.rand(50, 50) < 0.3).astype(np.uint8)
    
    result_2pass = label_cc_2pass(img)
    result_bfs = label_cc_bfs(img)
    result_dfs = label_cc_dfs(img)
    result_dsu = label_cc_dsu(img)
    
    n_2pass = len(np.unique(result_2pass[result_2pass > 0]))
    n_bfs = len(np.unique(result_bfs[result_bfs > 0]))
    n_dfs = len(np.unique(result_dfs[result_dfs > 0]))
    n_dsu = len(np.unique(result_dsu[result_dsu > 0]))
    
    assert n_2pass == n_bfs == n_dfs == n_dsu, \
        f"Component count mismatch: 2pass={n_2pass}, BFS={n_bfs}, DFS={n_dfs}, DSU={n_dsu}"
    
    print("✓ Consistency test passed")


def run_all_tests():
    """Run all tests"""
    print("Running tests...\n")
    try:
        test_simple_4_connected()
        test_simple_8_connected()
        test_empty_image()
        test_full_foreground()
        test_consistency()
        print("\n✅ All tests passed!")
        return 0
    except AssertionError as e:
        print(f"\n❌ Test failed: {e}")
        return 1
    except Exception as e:
        print(f"\n❌ Error: {e}")
        import traceback
        traceback.print_exc()
        return 1


if __name__ == "__main__":
    sys.exit(run_all_tests())

